File: csharp/src/Noie.Nds.Api/Adapter/ContextAdapter.cs
using Noie.Nds.Api.Context;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-CONTEXTADAPTER-000] Context adapter (INdsContext ↔ proto-compatible shapes).
/// 
/// <para>[Semantic] Conversion utilities for tracing/correlation context.</para>
/// </summary>
public static class ContextAdapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-CONTEXTADAPTER-010] Create a proto-compatible context tuple.
    /// </summary>
    public static (string TraceId, string CorrelationId, Dictionary<string, string> Meta) ToProtoData(INdsContext context)
    {
        return (
            context.TraceId,
            context.CorrelationId,
            new Dictionary<string, string>(context.Meta)
        );
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-CONTEXTADAPTER-011] Create a domain context from proto-compatible data.
    /// </summary>
    public static INdsContext FromProtoData(string traceId, string correlationId, IReadOnlyDictionary<string, string>? meta)
    {
        var context = NdsContext.Create(traceId, correlationId);
        if (meta != null && meta.Count > 0)
        {
            context = context.WithMeta(meta);
        }
        return context;
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-CONTEXTADAPTER-012] Create a domain context from proto-compatible data (Dictionary overload).
    /// </summary>
    public static INdsContext FromProtoData(string traceId, string correlationId, Dictionary<string, string>? meta)
    {
        return FromProtoData(traceId, correlationId, (IReadOnlyDictionary<string, string>?)meta);
    }
}

File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/ContextAdapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import noie.linmimeng.noiedigitalsystem.api.context.NdsContext;
import java.util.Map;
import java.util.HashMap;

/**
 * [Index] NDS-JAVA-CONTEXT-000
 * [Semantic] NdsContext Domain ↔ Proto conversion.
 *
 * @since 2.0.0
 */
public final class ContextAdapter {
    
    private ContextAdapter() {
        // [Index] NDS-JAVA-CONTEXT-001 [Constraint] Utility class; instantiation is prohibited.
    }
    
    /**
     * @param domain domain NdsContext (nullable)
     * @return Proto NdsContext; null if input is null
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.context.NdsContext toProto(NdsContext domain) {
        if (domain == null) {
            return null;
        }
        
        noie.linmimeng.noiedigitalsystem.api.proto.context.NdsContext.Builder builder = 
            noie.linmimeng.noiedigitalsystem.api.proto.context.NdsContext.newBuilder()
                .setTraceId(domain.traceId())
                .setCorrelationId(domain.correlationId())
                .setTimestamp(System.currentTimeMillis());
        
        if (domain.meta() != null) {
            builder.putAllMeta(domain.meta());
        }
        
        return builder.build();
    }
    
    /**
     * @param proto Proto NdsContext (nullable)
     * @return domain NdsContext; null if proto is null or traceId is empty
     */
    public static NdsContext fromProto(noie.linmimeng.noiedigitalsystem.api.proto.context.NdsContext proto) {
        if (proto == null || proto.getTraceId().isEmpty()) {
            return null;
        }
        
        String traceId = proto.getTraceId();
        String correlationId = proto.getCorrelationId().isEmpty() ? traceId : proto.getCorrelationId();
        Map<String, String> meta = new HashMap<>(proto.getMetaMap());
        
        // [Index] NDS-JAVA-CONTEXT-010 [Behavior] Construct via factory, then enrich with metadata.
        NdsContext context = NdsContext.create(traceId, correlationId);
        if (!meta.isEmpty()) {
            context = context.withMeta(meta);
        }
        return context;
    }
}
